File: examples/endpoint_internal_cache.cpp
#include "hakoniwa/pdu/endpoint.hpp"
#include <iostream>
#include <vector>

int main() {
    hakoniwa::pdu::Endpoint endpoint("internal_cache", HAKO_PDU_ENDPOINT_DIRECTION_INOUT);

    if (endpoint.open("config/sample/endpoint_internal_cache.json") != HAKO_PDU_ERR_OK) {
        std::cerr << "Failed to open internal cache endpoint" << std::endl;
        return 1;
    }
    if (endpoint.start() != HAKO_PDU_ERR_OK) {
        std::cerr << "Failed to start internal cache endpoint" << std::endl;
        return 1;
    }

    hakoniwa::pdu::PduResolvedKey key;
    key.robot = "TestRobot";
    key.channel_id = 1;

    std::vector<std::byte> write_data = {std::byte(0x01), std::byte(0x02)};
    if (endpoint.send(key, write_data) != HAKO_PDU_ERR_OK) {
        std::cerr << "Send failed" << std::endl;
        return 1;
    }

    std::vector<std::byte> read_buffer(16);
    size_t read_len = 0;
    if (endpoint.recv(key, read_buffer, read_len) != HAKO_PDU_ERR_OK) {
        std::cerr << "Recv failed" << std::endl;
        return 1;
    }

    std::cout << "Read back " << read_len << " bytes from internal cache" << std::endl;

    (void)endpoint.stop();
    (void)endpoint.close();
    return 0;
}

File: python/examples/endpoint_internal_cache.py
#!/usr/bin/env python3
import sys
from pathlib import Path


def main() -> int:
    repo_root = Path(__file__).resolve().parents[2]
    python_root = repo_root / "python"
    sys.path.insert(0, str(python_root))

    from hakoniwa_pdu_endpoint.c_endpoint import Endpoint, PduResolvedKey

    endpoint = Endpoint("py_example_internal_cache", "inout")
    endpoint.open(str(repo_root / "config/sample/endpoint_internal_cache.json"))
    endpoint.start()

    key = PduResolvedKey(robot="py_example_robot", channel_id=1)
    payload = b"\x01\x02\x03\x04"
    endpoint.send(key, payload)
    recv_payload = endpoint.recv(key, 16)

    print("received:", list(recv_payload))

    endpoint.stop()
    endpoint.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
